### DISCUSSION TITLE
How to apply this to a hexagonal board game?

### DISCUSSION THREAD

**Discussion by imsohandsome123** [suragnair/alpha-zero-general#204] [Q&A]
As title, I want to apply this to Abalone. 
I know how to apply this on a rectangle board game. 
However, the board of Abalone is hexagon shape.
Is there a recommended way to apply this to a hexagon board game or is there any exists example that I can refer to?

**Comment by Ettrig** [comment]
Hope you are a good programmer as well as handsome. I made a Hex player, which was suprisingly much work. My impression is that Abalone is much more complex. Good exercise though. 

I didn't find a good way to represent hexagonal boards. So I used a (common) two-dimensional matrix. Each "cell" in the matrix has 4 or 8 "natural" neighbors. I chose to recognize the 4 straight plus up-right and down-left from the oblique neighbors. Then I just had to follow this convention when coding the game functions. That is OK for Hex, but maybe too cumbersome for  Abalone. 

If you squint a little when looking at https://commons.wikimedia.org/wiki/File:Hexposition02.jpg you get the idea. At least I did. 

Good luck!

**Comment by imsohandsome123** [comment]
> Hope you are a good programmer as well as handsome. I made a Hex player, which was suprisingly much work. My impression is that Abalone is much more complex. Good exercise though.
> 
> I didn't find a good way to represent hexagonal boards. So I used a (common) two-dimensional matrix. Each "cell" in the matrix has 4 or 8 "natural" neighbors. I chose to recognize the 4 straight plus up-right and down-left from the oblique neighbors. Then I just had to follow this convention when coding the game functions. That is OK for Hex, but maybe too cumbersome for Abalone.
> 
> If you squint a little when looking at https://commons.wikimedia.org/wiki/File:Hexposition02.jpg you get the idea. At least I did.
> 
> Good luck!

Do you think padding is a good idea?

For example, this is the original board representation I used:
       [  -1, -1, 0, 1, 1  ], 
      [  -1, -1, -1, 1, 1, 1  ],
     [  0, -1, -1, 0, 1, 1, 0  ],
   [  0, 0, 0, 0, 0, 0, 0, 0  ],
  [  0, 0, 0, 0, 0, 0, 0, 0, 0 ],
   [  0, 0, 0, 0, 0, 0, 0, 0  ],
    [  0, 1, 1, 0, -1, -1, 0  ],
      [  1, 1, 1, -1, -1, -1  ],
       [  1, 1, 0, -1, -1  ]
Where ones stand for black, minus ones stand for white, and zero stand for empty empty squares.

Padding example:
[  2, 2, -1, -1, 0, 1, 1, 2, 2  ], 
[  2, 2, -1, -1, -1, 1, 1, 1, 2  ],
[  2, 0, -1, -1, 0, 1, 1, 0, 2  ],
[  2, 0, 0, 0, 0, 0, 0, 0, 0  ],
[  0, 0, 0, 0, 0, 0, 0, 0, 0  ],
[  0, 0, 0, 0, 0, 0, 0, 0, 2  ],
[  2, 0, 1, 1, 0, -1, -1, 0, 2  ],
[  2, 1, 1, 1, -1, -1, -1, 2, 2  ],
[  2, 2, 1, 1, 0, -1, -1, 2, 2  ]
Where ones stand for black, minus one stand for white, zero stand for empty empty squares, and twos stand for padding squares.

**Comment by Ettrig** [comment]
My advice would be not to spend much thought on that now. Yes, padding is probably necessary, because we want many convolutional layers and without padding they would shrink into non-existence. 

But my hunch is that if everything else is working correctly, any fill pattern will work. So if you start with zero padding, then you can easily experiment with other filling patterns later, to see they result in faster improvement or better end results.
